File: android/companion/bridge-core/src/main/kotlin/dev/picoswitch/bridge/protocol/BridgeOutputCodec.kt
package dev.picoswitch.bridge.protocol

import dev.picoswitch.bridge.core.BridgeOutput
import dev.picoswitch.bridge.core.RumbleRequest

/**
 * PicoSwitch Bridge output report -> normalized [BridgeOutput].
 *
 * The framing tolerance here is protocol, not a platform workaround: HID hosts
 * legitimately deliver an output report either on the interrupt channel or as a
 * control-channel SET_REPORT, and stacks differ on whether the report ID is
 * included in the payload. Both framings carry the same five bytes, so the
 * decoder accepts both and every backend gets the tolerance for free.
 *
 * Returns null when the payload cannot be a bridge output report, so a stray
 * report can never be applied as rumble.
 */
object BridgeOutputCodec {
    /** `[id][rumble L][rumble R][player][flags]` */
    const val REPORT_SIZE_WITH_ID = 5
    const val BODY_SIZE = 4

    const val FLAG_MOTION_WANTED = 0x01

    fun decode(data: ByteArray?, reportId: Int? = null): BridgeOutput? {
        if (data == null) return null
        if (reportId != null && reportId != ControllerReportEncoder.OUTPUT_REPORT_ID) return null
        var body = data
        // Tolerate an embedded report ID: some stacks include it in `data`.
        if (body.size == REPORT_SIZE_WITH_ID &&
            body[0].toInt() and 0xFF == ControllerReportEncoder.OUTPUT_REPORT_ID
        ) {
            body = body.copyOfRange(1, body.size)
        }
        if (body.size < BODY_SIZE) return null
        return BridgeOutput(
            rumble = RumbleRequest(
                left = body[0].toInt() and 0xFF,
                right = body[1].toInt() and 0xFF,
            ),
            playerIndicator = body[2].toInt() and 0xFF,
            motionRequested = (body[3].toInt() and FLAG_MOTION_WANTED) != 0,
        )
    }
}

File: windows/companion/src/PicoSwitch.Bridge.Core/Protocol/BridgeOutputCodec.cs
using PicoSwitch.Bridge.Core;

namespace PicoSwitch.Bridge.Protocol;

/// <summary>
/// PicoSwitch Bridge output report -> normalized <see cref="BridgeOutput"/>.
///
/// The framing tolerance here is protocol, not a platform workaround: HID hosts
/// legitimately deliver an output report either on the interrupt channel or as a
/// control-channel SET_REPORT, and stacks differ on whether the report ID is
/// included in the payload. Both framings carry the same five bytes, so the
/// decoder accepts both and every backend gets the tolerance for free.
///
/// Returns null when the payload cannot be a bridge output report, so a stray
/// report can never be applied as rumble.
/// </summary>
public static class BridgeOutputCodec
{
    /// <summary><c>[id][rumble L][rumble R][player][flags]</c></summary>
    public const int ReportSizeWithId = 5;

    public const int BodySize = 4;

    public const int FlagMotionWanted = 0x01;

    public static BridgeOutput? Decode(byte[]? data, int? reportId = null)
    {
        if (data is null)
        {
            return null;
        }

        if (reportId is not null && reportId != ControllerReportEncoder.OutputReportId)
        {
            return null;
        }

        var body = new ReadOnlySpan<byte>(data);

        // Tolerate an embedded report ID: some stacks include it in `data`.
        if (body.Length == ReportSizeWithId && body[0] == ControllerReportEncoder.OutputReportId)
        {
            body = body[1..];
        }

        if (body.Length < BodySize)
        {
            return null;
        }

        return new BridgeOutput(
            Rumble: new RumbleRequest(Left: body[0], Right: body[1]),
            PlayerIndicator: body[2],
            MotionRequested: (body[3] & FlagMotionWanted) != 0);
    }
}
